db/snapshot/backup_task.cc

```diff
@@ -163,16 +166,56 @@ future<> backup_task_impl::backup_sstable(sstables::generation_type gen, const c
     }
 };
 
+void backup_task_impl::on_unlink(sstables::generation_type gen) {
+    if (_sstable_comps.contains(gen)) {
+        _unlinked_sstables.push_back(gen);
+    }
+}
+
 future<> backup_task_impl::do_backup() {
     if (!co_await file_exists(_snapshot_dir.native())) {
         throw std::invalid_argument(fmt::format("snapshot does not exist at {}", _snapshot_dir.native()));
     }
 
+    _backup_shard = this_shard_id();
+
+    co_await _subscriptions.start();
+    co_await _snap_ctl.db().invoke_on_all([&] (replica::database& db) {
+        auto schema = db.find_schema(_table_id);
+        auto sub = db.get_sstables_manager(*schema).subscribe([&] (sstables::generation_type gen, sstables::manager_event_type event) -> future<> {
+            switch (event) {
+            case sstables::manager_event_type::add:
+                break;
+            case sstables::manager_event_type::unlink:
+                if (_sstable_comps.contains(gen)) {
```

Comment:
Isn't it racy? The _sstable_comps lives on "_backup_shard" but is checked on whatever shard the notification comes to

Reply:
It's checked again on _backup_shard in backup_task_impl::on_unlink.
This check here just eliminates the cross-shard task which is going to be the common case (since the notifications are global, for all tables)

Reply:
I mean another race -- if one shard updates the `_sstable_comps` and another one checks it, won't they step of use-after-free or not-self-consistent internals of the collection?

Reply:
I don't think so since `contains` is read-only and `extract` does not invalidate iterators so it won't reallocate the vector of buckets.

Reply:
I will add an immutable set of generation in snapshot, just in case.

Reply:
checking an immutable set of sstable generations in snapshot in v3 ([7ab3391](https://github.com/scylladb/scylladb/commit/7ab3391f15e68adfff6187385be438b96bb3d57e))